### PR TITLE
Fix app generation for the Samourai app

### PR DESCRIPTION
**Pull request by AaronDewes**
This also only loads the schema file for validation once instead for every app, which should improve the performance of the app validation

### CODE REVIEW

File: app/lib/composegenerator/v3/generate.py [hunk 1]
```diff
@@ -85,11 +85,12 @@ def createComposeConfigFromV3(app: dict, nodeRoot: str):
...
     # This is validated earlier
     for container in newApp.containers:
-        container.ports = container.requiredPorts
+        for tcpPort in container.requiredPorts:
+            container.ports.append("{}:{}".format(tcpPort, tcpPort))
         del container.requiredPorts
     for container in newApp.containers:
         for udpPort in container.requiredUdpPorts:
-            container.ports.append("{}/udp".format(udpPort))
+            container.ports.append("{}/udp:{}/udp".format(udpPort, udpPort))
```
Comment (AaronDewes) [inline app/lib/composegenerator/v3/generate.py:93]: To document these changes somewhat, I thought if the port on the inside and outside was the same, directly having the port in the ports section without a target port would be enough, but it seems like docker needs this explicit definition.

File: app/lib/composegenerator/v2/utils/networking.py [hunk 2]
```diff
@@ -77,14 +77,14 @@ def getContainerHiddenService(
...
-                otherHiddenServices += "\n"
                 if isinstance(value, int):
                     otherHiddenServices += "# {} {} {} Hidden Service\nHiddenServiceDir /var/lib/tor/app-{}-{}\n".format(
-                        metadata.name, container.name, key, metadata.id, container.name
+                        metadata.name, container.name, key, metadata.id, key
                     )
                     otherHiddenServices += "HiddenServicePort {} {}:{}".format(
                         value, containerIp, value
                     )
+                    otherHiddenServices += "\n"
```
Comment (AaronDewes) [inline app/lib/composegenerator/v2/utils/networking.py:87]: This was moved (from line 80) to make the generated torrc files prettier and does not really fix a bug.

File: app/lib/composegenerator/v2/utils/networking.py [hunk 3]
```diff
@@ -77,14 +77,14 @@ def getContainerHiddenService(
                 else:
                     additionalHiddenServices[key] = value
             for key, value in additionalHiddenServices.items():
-                otherHiddenServices += "\n"
                 if isinstance(value, int):
                     otherHiddenServices += "# {} {} {} Hidden Service\nHiddenServiceDir /var/lib/tor/app-{}-{}\n".format(
-                        metadata.name, container.name, key, metadata.id, container.name
+                        metadata.name, container.name, key, metadata.id, key
```
Comment (AaronDewes) [inline app/lib/composegenerator/v2/utils/networking.py:82]: This fixes the names for the generated hidden service directories.

File: app/lib/composegenerator/v2/utils/networking.py [hunk 4]
```diff
@@ -47,15 +47,15 @@ def getContainerHiddenService(
         if isinstance(container.hiddenServicePorts, int):
             return getHiddenServiceString(
                 "{} {}".format(metadata.name, container.name),
-                metadata.id,
+                metadata.id if isMainContainer else "{}-{}".format(metadata.id, container.name),
```
Comment (AaronDewes) [inline app/lib/composegenerator/v2/utils/networking.py:50]: This adds the container name if the container is not the main container to ensure there are no conflicts between containers

File: app/lib/validate.py [hunk 5]
```diff
@@ -7,18 +7,18 @@
...
 import yaml
 
+scriptDir = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..")
+
+with open(os.path.join(scriptDir, 'app-standard-v1.yml'), 'r') as f:
+    schemaVersion1 = yaml.safe_load(f)
+with open(os.path.join(scriptDir, 'app-standard-v2.yml'), 'r') as f:
+    schemaVersion2 = yaml.safe_load(f)
+with open(os.path.join(scriptDir, 'app-standard-v3.yml'), 'r') as f:
+    schemaVersion3 = yaml.safe_load(f)
```
Comment (AaronDewes) [inline app/lib/validate.py:17]: This ensures schema files are only loaded once

**Comment by pwltr** [comment]
Thanks for the annotations, clears things up! How do I best apply this and test? I checked out this branch, ran `sudo ./scripts/app generate` (+ download, update) and installed Samourai. Got the following error:

```
citadel@citadel-dev:~/citadel$ docker logs -f samourai-server-main-1
2022/06/03 18:58:18 [crit] 65#65: pread() "/etc/nginx/sites-enabled/dojo.conf" failed (21: Is a directory)
nginx: [crit] pread() "/etc/nginx/sites-enabled/dojo.conf" failed (21: Is a directory)
```

**Comment by AaronDewes** [comment]
Can you try uninstalling the app, checking out this branch, running the start script, then installing again? It seems like the app didn't get copied into app-data properly, so please also make sure to `rm -rf app-data/samourai-server` after the uninstall.

I could not reproduce this in my testing env, so I'm not fully sure

**Review by pwltr — APPROVED** [review]

**Review by WilliamConnatser — APPROVED** [review]
Tested and working in my dev environment. Nice work! :)

**Pull request merged**